pkg/cloud/services/networking/securitygroups.go

```diff
@@ -157,19 +208,70 @@ func (s *Service) generateDesiredSecGroups(openStackCluster *infrav1.OpenStackCl
 		}
 	}
 
-	desiredSecGroups[controlPlaneSuffix] = infrav1.SecurityGroup{
+	desiredSecGroups[controlPlaneSuffix] = securityGroupSpec{
 		Name:  secGroupNames[controlPlaneSuffix],
 		Rules: controlPlaneRules,
 	}
 
-	desiredSecGroups[workerSuffix] = infrav1.SecurityGroup{
+	desiredSecGroups[workerSuffix] = securityGroupSpec{
 		Name:  secGroupNames[workerSuffix],
 		Rules: workerRules,
 	}
-
 	return desiredSecGroups, nil
 }
 
+// getAllNodesRules returns the rules for the allNodes security group that should be created.
+func getAllNodesRules(remoteManagedGroups map[string]string, allNodesSecurityGroupRules []infrav1.SecurityGroupRuleSpec) ([]resolvedSecurityGroupRuleSpec, error) {
+	rules := make([]resolvedSecurityGroupRuleSpec, 0, len(allNodesSecurityGroupRules))
+	for _, rule := range allNodesSecurityGroupRules {
+		if err := validateRemoteManagedGroups(remoteManagedGroups, rule.RemoteManagedGroups); err != nil {
+			return nil, err
+		}
+		r := resolvedSecurityGroupRuleSpec{
+			Description:    rule.Description,
+			Direction:      rule.Direction,
+			EtherType:      rule.EtherType,
+			PortRangeMin:   rule.PortRangeMin,
+			PortRangeMax:   rule.PortRangeMax,
+			Protocol:       rule.Protocol,
+			RemoteGroupID:  rule.RemoteGroupID,
+			RemoteIPPrefix: rule.RemoteIPPrefix,
+		}
+
+		if len(rule.RemoteManagedGroups) > 0 {
+			if rule.RemoteGroupID != "" {
+				return nil, fmt.Errorf("remoteGroupID must not be set when remoteManagedGroups is set")
+			}
+
+			for _, rg := range rule.RemoteManagedGroups {
+				if _, ok := remoteManagedGroups[rg.String()]; !ok {
+					return nil, fmt.Errorf("remoteManagedGroups: %s is not a valid remote managed security group", rg)
+				}
```

Comment:
Aren't you already doing these validations in `validateRemoteManagedGroups()`?

Reply:
good spot